### PR TITLE
Add testing framework skeleton & several initial tests

### LINKED ISSUE #3162: Integrate chosen testing framework into the dapp

### PR DESCRIPTION
**Pull request by chinins**
## Description

This PR installs cypress & other helper libraries. It adds initial cypress settings and provides explanation on how to run/operate cypress. It includes the cypress run/open scripts. It adds a couple of custom reusable commands and a couple of example tests.

I have not cleaned up the mint tokens action selectors on purpose - just to show the options that we have. There is a separate issue for that.

*Please read the whole description.* I intend to have this PR as a source of intro info, useful links, design decision/discussions. I will periodically update it.

Contents:
- difficulties of testing the web3 apps & my solution proposals.
- discussion points
- How to install and run
- Useful links

**DApp complications**
The tricky part with the dapp is that it's slow and you need to wait for some operations for a long time. Hence it takes a long time to run tests. Usually when you test you run all of your tests but I think that it's not feasible for our dapp. E.g., you won't create a colony or claim a new username every time (unless you are working specifically on that part and want to test your work). So I think that it's worth to run this only once at the start and after that you can skip these tests (there are `skip` and `only` commands to do that)

I spent a lot of time researching for how we could wait for a certain UI element to appear (e.g. the green tick icon that marks the success of the operation) and it seems it's not really possible. So for now I used a combination of several methods but for all of them it's just increase the wait time in different ways. Here is a good blog post about different options for `waiting` & its alternatives that's clear and not too long:

https://filiphric.com/waiting-in-cypress-and-how-to-avoid-it?utm_campaign=Software%2BTesting%2BWeekly&utm_medium=email&utm_source=Software_Testing_Weekly_106

I have also experimented with waiting for API calls that we do to `http://localhost:8545` but we can't distinguish between the calls. I discussed it with Raul and we agreed that different delay/wait options is the way now. Please, let me know if you have any other ideas.

Potentially when we have Raul's version of the dapp we could use cy.request  & cy.intercept to avoid `wait` & `timeout` options.

Discussion questions:

- is that ok to run some tests only once at the beginning of work (e.g. creating colony, claiming user, installing extensions) and then just use it in the future tests (unless of course a certain part is the focus of work)?
- is it ok to do different combinations of waiting time?
- any other suggestions/remark you have for the testing part?

**Install/Run tests**
1. `npm i` as we have new dev-dependencies

2. You have to have the dev environment running with webpack running and you have to have a colony created (until we have a test that creates a colony).

3. In `cypress.json` in the root directory you need to update the name & the nativeToken symbol of your colony.

4. Running `npm i` will also install cypress interactive GUI that looks like below. To open it you can either run `npx cypress open` or  `npm run cy:open`. To run the tests in the browser you just click `Run integration tests` in this cypress GUI.
Screenshot 2022-02-18 at 16 55 14

5. To run cypress tests in a headless mode you run `npx cypress run` or `npm run cy:run`.

Useful links:
- cypress docs:
https://docs.cypress.io/guides/getting-started/installing-cypress

- best practices
https://docs.cypress.io/guides/references/best-practices

- a good blog that covers cypress in a more user friendly way
https://filiphric.com/blog

resolves #3162

### CODE REVIEW

**Review by dit7ya — COMMENTED** [review]
The colony name is hardcoded at a couple of places - so I had to create a colony matching those values instead of modifying the `cypress.json` file.

Clicking `New Action` failed since it was disabled due to (I am guessing) the user that logs in does `Join` the colony and hence the `New Action` button is disabled.




Also, I am curious about your thoughts on why you chose Cypress over Playwright if we are doing E2E tests anyway.

---
**Ignore the following safely**

I know no one else in the team uses NixOS but I am writing this here for reference in case someone does in future.

Getting Cypress to work in NixOS is nontrivial. In my `flake.nix` I had to add the correct revision of [Cypress](https://search.nixos.org/packages?channel=unstable&from=0&size=50&sort=relevance&type=packages&query=cypress) according to this project and have to call it with 
```fish
CYPRESS_RUN_BINARY=(which Cypress) npx cypress open
```

File: cypress/e2e/2-actions.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  'Colony can mint tokens via aciton',
+  { defaultCommandTimeout: 10000 },
+  () => {
+    it('mint native tokens', () => {
+      const amountToMint = 10;
+      const annotationText = 'Test annotation';
+
+      cy.login();
+
+      cy.visit('/colony/a');
```
[the remark this answers is not in the document]
Comment (chinins) [reply cypress/e2e/2-actions.js:11]: yes, thank you! I'll update.

**Comment by chinins** [comment]
> The colony name is hardcoded at a couple of places - so I had to create a colony matching those values instead of modifying the `cypress.json` file.
> 
> Clicking `New Action` failed since it was disabled due to (I am guessing) the user that logs in does `Join` the colony and hence the `New Action` button is disabled.
> 
> 
> 
> Also, I am curious about your thoughts on why you chose Cypress over something like [Playwright](https://playwright.dev/) if we are doing E2E tests anyway.
> 
> **Ignore the following safely**
> 
> I know no one else in the team uses NixOS but I am writing this here for reference in case someone does in future.
> 
> Getting Cypress to work in NixOS is nontrivial. In my `flake.nix` I had to add the correct revision of [Cypress](https://search.nixos.org/packages?channel=unstable&from=0&size=50&sort=relevance&type=packages&query=cypress) according to this project and have to call it with
> 
> ```fish
> CYPRESS_RUN_BINARY=(which Cypress) npx cypress open
> ```

It was disabled because the extensions data was not loaded yet as you can see from your print screen. So it just probably needs more time... I assume because our systems are different the processing time is loger on yours...

As for the join, did you create a colony when you were logged in as the first wallet address in Ganache? I should make that clear in my description as well and in the future it will be automated with creating colony tests anyway.

File: cypress/e2e/2-actions.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (chinins) [reply cypress/e2e/2-actions.js]: fixed in the left places where it was hardcoded

**Comment by dit7ya** [comment]
I guess another reason for choosing Playwright is that it supports auto-wait. Writing arbitrary wait times has the potential to cause a lot of trouble in the long run, IMO.

> Auto-wait. Playwright waits for elements to be actionable prior to performing actions. It also has a rich set of introspection events. The combination of the two eliminates the need for artificial timeouts - the primary cause of flaky tests.

File: cypress/e2e/1-logins.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+      cy.visit('/')
+        .contains(/connect wallet/i)
+        .click();
+
+      // eslint-disable-next-line cypress/no-unnecessary-waiting
+      cy.contains(/ganache/i)
+        .click()
+        // click on drowpdown with accounts
+        .get('.Select_activeOption_1sMcrZHH > span')
```
Comment (alicjakujawa) [inline cypress/e2e/1-logins.js:25]: I would avoid any "getting elements by classes". Please read this doc: https://docs.cypress.io/guides/references/best-practices

Comment (chinins) [reply cypress/e2e/1-logins.js:25]: As I wrote in my PR description:
"I have not cleaned up the mint tokens action selectors on purpose - just to show the options that we have. There is a separate issue for that." And there is this link in the PR description as well.

**Comment by arrenv** [comment]
I managed to get Cypress working in my dev environment. So, all looks good from here :)

**Review by ArmandoGraterol — APPROVED** [review]
Looks good! 

I would argue that every single element that is being selected in the commands and tests should be selected by using `data-test`, but then again the elements that are being selected are mostly, if not all, core components so their style/text most likely won't change

File: cypress/e2e/1-logins.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+describe('Claim new user name', () => {
+  if (!Cypress.config().skipInitTests) {
+    it('logs in new user', () => {
+      /*
+        this will be the first address in the list;
+        can be changed to further order in the list if the test needs to be repeated
+      */
+      cy.claimNewUserName(0);
+
+      cy.url({ timeout: 10000 }).should(
```
[the remark this answers is not in the document]
Comment (chinins) [reply cypress/e2e/1-logins.js:19]: good catch! I thought I removed it everywhere.

**Comment by chinins** [comment]
> Looks good!
> 
> I would argue that every single element that is being selected in the commands and tests should be selected by using `data-test`, but then again the elements that are being selected are mostly, if not all, core components so their style/text most likely won't change

Yeah, I agree with the `data-test` selector point. I have updated it now. The only one not like this left is to select an option from a list. I quite like it this way as we can change the number chosen in the dropdown list of addresses very easily.

**Pull request merged**